libs/pbd/pbd/controllable.h

```diff
@@ -147,7 +147,7 @@ class LIBPBD_API Controllable : public PBD::StatefulDestructible, public std::en
 	static PBD::Signal1<void, std::weak_ptr<PBD::Controllable> > GUIFocusChanged;
 	static PBD::Signal1<void, std::weak_ptr<PBD::Controllable> > ControlTouched;
 
-	PBD::Signal2<void,bool,PBD::Controllable::GroupControlDisposition> Changed;
+	PBD::Signal3<void,bool,PBD::Controllable::GroupControlDisposition,boost::optional<double>> Changed;
```

Comment:
I am not really a fan of adding the the value to the Changed signal here. That is a huge commitment, particularly since it is unused except for VST3, and  basically a workaround for yabrdge/VST3 only.

Do you think we could instead cache that value on VST3Plugin  or the implementation VST3PI for a quick lookup instead?

Reply:
Hey Robin! Thanks for your answer.

Sure, I could try another approach, I thought that this was the best way because it seems logic to me that the value created with the event goes with the signal.

Regarding caching the value, how does it come to your mind?
One way could be using a sort of stack or queue, but how could it be mantained? Delete the value after one read? What would happend to the other calls to "get_value" caused by the same event? They will lose the chance to get a cached value.

Another way could be caching the value with a timestamp, but I'm thinking that it could be too complex for something like this.

Or caching it during the performEdit until the next beginEdit, but maybe there is a time when we need genuinely fetch the actual value through "get_value" and I could get a cached old one.

What do you think?

Also, which are the cons for passing the value in the signal?

Regards

Reply:
> Also, which are the cons for passing the value in the signal?

The purpose of the signal is to just inform others that something has changed. That is semantically different from passing a ValueChange.  Also, when an optional value is passed, all API users (notably control surfaces) will need to check if its boost::none, and provide a fall back.  Then the value may change more rapidly than a given UI can update.

> Regarding caching the value, how does it come to your mind?

The basic idea I had is to cache the normalized value at the point when the `Changed()` signal is emitted (here `VST3PI::ParamValueChanged`).   Then return the last cache value in `VST3PI::get_parameter`.

As for the the implementation my first thought that for every call 
   `_shadow_data[p] = value;`
we can add 
 ` _shadow_data_plain[p] = normalizedParamToPlain (id, _shadow_data[p]);`

--

Then again for other plugins (some have 1000+ controls) this approach may add a bit of load if a given value is never used.

Reply:
Sorry for not replying sooner. Somehow this dropped below my radar.

So just change the VST3 `OnParameterChange` signal to pass along the value (and cache it)?
That sounds reasonable.